src/modm/platform/clock/sam_pmc/clockgen_impl.hpp.in

```diff
@@ -11,34 +11,89 @@
 
 namespace modm::platform {
 
-template <MasterClkSource src, MasterClkPrescaler pres>
+template <MainInternalFreq freq>
+void ClockGen::enableMainInternal() {
+	// Enable Main RC Oscillator and wait for it to stabilize
+	PMC->CKGR_MOR |= CKGR_MOR_KEY_PASSWD | CKGR_MOR_MOSCRCEN;
+	while (!(PMC->PMC_SR & PMC_SR_MOSCRCS));
+
+	// Change Main RC Oscillator frequency
+	PMC->CKGR_MOR =
+		(PMC->CKGR_MOR & ~(CKGR_MOR_MOSCRCF_Msk)) |
+		CKGR_MOR_MOSCRCF((uint32_t)freq) |
+		CKGR_MOR_KEY_PASSWD;
+	while (!(PMC->PMC_SR & PMC_SR_MOSCRCS));
+}
+
+template <MainClockSource src>
+void ClockGen::selectMainClockSource() {
+	if (src == MainClockSource::Internal) {
+		PMC->CKGR_MOR =
+			(PMC->CKGR_MOR & ~CKGR_MOR_MOSCSEL) |
+			CKGR_MOR_KEY_PASSWD;
+	} else {  // MainClockSource::External
+		PMC->CKGR_MOR |= CKGR_MOR_MOSCSEL | CKGR_MOR_KEY_PASSWD;
+	}
+}
+
+template <MasterClkSource src, MasterClkPrescaler pres, MasterClkDivider divider>
 void ClockGen::selectMasterClk() {
 	// Datasheet says when selecting PLL as source, write PRES first, otherwise
 	// write CSS first.
-	if(src == MasterClkSource::PLLA_CLK || src == MasterClkSource::PLLB_CLK) {
+%% if target["family"] in ["g"]
+	static_assert(divider == MasterClkDivider::Div1, "Divider for master clock is not supported");
```

Comment:
Yeah, I can kinda see the usefulness, but the validation of clock configs is hard to do without having a full system view (at least that was our experience on STM32). I would delegate this to the calling function or even the user.

Reply:
I'd prefer to leave it as-is for this PR. I already dived into this rabbit hole much deeper than I anticipated. And I have a lot of other rabbit holes to explore.

Reply:
Yeah I'm not sure what you're wanting to change Niklas. On samg, this divider doesnt exist, so it makes sense to me to assert it's not provided.

Reply:
No problem, we'll refactor it later. The important thing is that it works.